### PR TITLE
feat: add meta project dependents command

### PR DESCRIPTION
**Pull request by mateodelnorte**
## Summary

- Adds `meta project dependents <name>` command that reads the `depends_on`/`provides` graph from `.meta.yaml`
- Lists all projects that directly depend on the given project
- Supports `--json` output for CI consumption
- Resolves `provides` tokens (e.g., `loop_lib` provides `loop-lib`, and `meta_cli` depends on `loop-lib`)

## Examples

```
$ meta project dependents meta_git_lib
meta_git_cli
meta_project_cli

$ meta project dependents loop_lib
loop_cli
meta_cli
meta_git_cli

$ meta project dependents meta_git_lib --json
["meta_git_cli","meta_project_cli"]
```

## Test plan

- [x] 6 new unit tests for `find_dependents` and the command itself
- [x] All 21 tests pass
- [x] Clippy clean with `-D warnings`
- [x] Manual testing against live `.meta.yaml`

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added a new `project dependents` command to list projects that depend on a specified project; supports recursive root discovery, token-matching (name/aliases), and returns lexicographically sorted results.
  * Outputs as newline-delimited or JSON; trailing `--json` forces JSON.

* **Documentation**
  * Help text updated with usage examples and option descriptions for the new command.

* **Tests**
  * Unit and integration tests added for matching rules, edge cases, resolution from subdirectories, and output modes.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -324,6 +329,100 @@ fn format_project_tree(nodes: &[ProjectTreeNode], output: &mut String, prefix: &
...
+        };
+        CommandResult::Message(json)
+    } else {
+        if dependents.is_empty() {
+            CommandResult::Message(format!("No projects depend on '{project_name}'."))
+        } else {
+            CommandResult::Message(dependents.join("\n"))
+        }
+    }
+}
```
[the remark this answers is not in the document]
Comment (mateodelnorte) [reply src/lib.rs:422]: Fixed.

File: src/lib.rs [hunk 2]
```diff
@@ -324,6 +329,100 @@ fn format_project_tree(nodes: &[ProjectTreeNode], output: &mut String, prefix: &
...
+    cwd: &Path,
+) -> CommandResult {
+    let project_name = match args.first() {
+        Some(name) => name.clone(),
+        None => {
+            return CommandResult::ShowHelp(Some(
+                "Usage: meta project dependents <project-name>".to_string(),
+            ))
+        }
+    };
```
[the remark this answers is not in the document]
Comment (mateodelnorte) [reply src/lib.rs:383]: Fixed.

File: src/lib.rs [hunk 3]
```diff
@@ -324,6 +329,100 @@ fn format_project_tree(nodes: &[ProjectTreeNode], output: &mut String, prefix: &
...
+        Err(e) => return CommandResult::Error(format!("Failed to parse meta config: {e}")),
+    };
+
+    let dependents = find_dependents(&project_name, &all_projects);
+
+    if options.json_output {
+        let json = match serde_json::to_string_pretty(&dependents) {
+            Ok(j) => j,
+            Err(e) => return CommandResult::Error(format!("Failed to serialize JSON: {e}")),
+        };
```
[the remark this answers is not in the document]
Comment (mateodelnorte) [reply src/lib.rs]: Fixed.

File: src/lib.rs [hunk 4]
```diff
@@ -72,6 +72,11 @@ pub fn execute_command(
         return CommandResult::ShowHelp(None);
     }
 
+    // project dependents reads the dependency graph directly
+    if command == "project dependents" {
+        return handle_project_dependents(args, options, cwd);
+    }
```
[the remark this answers is not in the document]
Comment (mateodelnorte) [reply src/lib.rs:78]: Fixed.

File: src/lib.rs [hunk 5]
```diff
@@ -324,6 +329,100 @@ fn format_project_tree(nodes: &[ProjectTreeNode], output: &mut String, prefix: &
...
+    let Some((meta_path, _format)) = config::find_meta_config_in(&start_dir) else {
+        return CommandResult::Error(format!("No .meta config found in {}", start_dir.display()));
+    };
+
+    let (all_projects, _ignore) = match config::parse_meta_config(&meta_path) {
+        Ok(result) => result,
+        Err(e) => return CommandResult::Error(format!("Failed to parse meta config: {e}")),
+    };
+
+    let dependents = find_dependents(&project_name, &all_projects);
```
[the remark this answers is not in the document]
Comment (mateodelnorte) [reply src/lib.rs]: Fixed.

File: src/lib.rs [hunk 6]
```diff
@@ -324,6 +329,100 @@ fn format_project_tree(nodes: &[ProjectTreeNode], output: &mut String, prefix: &
...
+
+    // Find all projects whose depends_on intersects with provided_tokens
+    let mut dependents: Vec<String> = all_projects
+        .iter()
+        .filter(|p| p.name != project_name)
+        .filter(|p| {
+            p.depends_on
+                .iter()
+                .any(|dep| provided_tokens.contains(dep.as_str()))
+        })
```
[the remark this answers is not in the document]
Comment (mateodelnorte) [reply src/lib.rs]: Fixed.

File: src/lib.rs [hunk 7]
```diff
@@ -324,6 +329,112 @@ fn format_project_tree(nodes: &[ProjectTreeNode], output: &mut String, prefix: &
...
+    };
+
+    let Some((meta_path, _format)) = config::find_meta_config_in(&start_dir) else {
+        return CommandResult::Error(format!("No .meta config found in {}", start_dir.display()));
+    };
+
+    let (all_projects, _ignore) = match config::parse_meta_config(&meta_path) {
+        Ok(result) => result,
+        Err(e) => return CommandResult::Error(format!("Failed to parse meta config: {e}")),
+    };
```
[the remark this answers is not in the document]
Comment (mateodelnorte) [reply src/lib.rs:403]: Valid observation. Currently `dependents` only reads the single nearest `.meta` config, which works for flat workspaces like the meta-repo itself. Recursive aggregation across nested `.meta` files (using `walk_meta_tree`) is a future enhancement — tracked in the broader `meta project` roadmap. Not blocking for the immediate CI use case.

**Comment by mateodelnorte** [comment]
🤖 **Agent starting** (`run-pr-review-1774469672`): `pr-review` run dispatched.

File: src/lib.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (mateodelnorte) [reply src/lib.rs:422]: Verified: already fixed in a12c942 — `json_from_args` extraction and `effective_options` pattern are in place at lines 345-355.

File: src/lib.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (mateodelnorte) [reply src/lib.rs:383]: Verified: already fixed in a12c942 — `args.iter().find(|a| !a.starts_with('-'))` is in place at line 357.

File: src/lib.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (mateodelnorte) [reply src/lib.rs]: Fixed in fd93f0d — added `normalize_token()` that folds hyphens→underscores and lowercases before comparison. Added test `test_find_dependents_normalized_tokens` covering cross-format matching.

File: src/lib.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (mateodelnorte) [reply src/lib.rs]: Acknowledged — recursive aggregation of nested `.meta` files is deferred as a design choice for flat workspaces, per author discussion in T7.

File: src/lib.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (mateodelnorte) [reply src/lib.rs:403]: Acknowledged — author has confirmed this is a reasonable scope boundary for flat workspaces. Resolving per the discussion above.

**Comment by mateodelnorte** [comment]
## Review Summary

| Commit | Comment | Fix |
|--------|---------|-----|
| fd93f0d | T6: Normalize dependency tokens before matching (@coderabbitai[bot]) | Added `normalize_token()` — folds hyphens→underscores + lowercases. Added test for cross-format matching. |

### Already fixed (verified & resolved)
- **T1** (`--json` extraction from args) — fixed in a12c942 (@greptile-apps[bot])
- **T2** (flag-before-positional misparse) — fixed in a12c942 (@greptile-apps[bot])

### Deferred by design
- **T5/T7** (recursive mode aggregating nested `.meta` files) — author confirmed flat-workspace scope is intentional (@coderabbitai[bot])

### Build status
- **clippy**: clean (0 warnings)
- **tests**: 22/22 passing

### Confidence: 95%
The only gap is recursive nested `.meta` aggregation (T5/T7), which is explicitly deferred as a design choice.

**Comment by mateodelnorte** [comment]
🤖 **Agent starting** (`run-pr-review-1774470949`): `pr-review` run dispatched.

**Comment by mateodelnorte** [comment]
## Review Fix Summary

| Commit | Comment | Fix |
|--------|---------|-----|
| 62c984f | T8: Climb to nearest ancestor `.meta` before loading dependents (@coderabbitai[bot]) | Always call `find_meta_config(cwd)` first so the command works from subdirectories within a flat workspace |
| 62c984f | T9: Resolve the target project before collecting its aliases (@coderabbitai[bot]) | Match target by normalized name or provides alias, then collect all its tokens |
| cc1e005 | Test coverage | Added `test_find_dependents_by_alias_query` and `test_project_dependents_from_subdirectory` |

**Build status**: clippy clean, 24/24 tests passing (including 2 new tests)
**Confidence**: 100%

🤖 Generated with [Claude Code](https://claude.com/claude-code)

File: src/lib.rs [hunk 8]
```diff
@@ -324,6 +329,122 @@ fn format_project_tree(nodes: &[ProjectTreeNode], output: &mut String, prefix: &
...
+                config::find_root_meta_dir(&nearest)
+            }
+            None => cwd.to_path_buf(),
+        }
+    } else {
+        cwd.to_path_buf()
+    };
+
+    let Some((meta_path, _format)) = config::find_meta_config_in(&start_dir) else {
+        return CommandResult::Error(format!("No .meta config found in {}", start_dir.display()));
```
[the remark this answers is not in the document]
Comment (mateodelnorte) [reply src/lib.rs:397]: Fixed in 62c984f — now always climbs to nearest ancestor `.meta` via `find_meta_config(cwd)` before loading dependents, so the command works from any subdirectory within a flat workspace. Added `test_project_dependents_from_subdirectory` to cover this.

File: src/lib.rs [hunk 9]
```diff
@@ -324,6 +329,122 @@ fn format_project_tree(nodes: &[ProjectTreeNode], output: &mut String, prefix: &
...
+        .filter(|p| {
+            p.depends_on
+                .iter()
+                .any(|dep| provided_tokens.contains(&normalize_token(dep)))
+        })
+        .map(|p| p.name.clone())
+        .collect();
+
+    dependents.sort();
+    dependents
```
[the remark this answers is not in the document]
Comment (mateodelnorte) [reply src/lib.rs]: Fixed in 62c984f — `find_dependents` now resolves the target by normalized name or provides alias before collecting tokens. Added `test_find_dependents_by_alias_query` to verify querying by alias finds all dependents.

**Comment by mateodelnorte** [comment]
🤖 **Agent starting** (`run-pr-review-1774471749`): `pr-review` run dispatched.

File: src/lib.rs [hunk 10]
```diff
@@ -324,6 +329,128 @@ fn format_project_tree(nodes: &[ProjectTreeNode], output: &mut String, prefix: &
...
+fn find_dependents(project_name: &str, all_projects: &[ProjectInfo]) -> Vec<String> {
+    // Resolve the target project by normalized name or alias
+    let normalized_query = normalize_token(project_name);
+    let Some(target) = all_projects.iter().find(|project| {
+        normalize_token(&project.name) == normalized_query
+            || project
+                .provides
+                .iter()
+                .any(|token| normalize_token(token) == normalized_query)
+    }) else {
```
[the remark this answers is not in the document]
Comment (mateodelnorte) [reply src/lib.rs]: Fixed in 9950cbd. Applied the suggested two-pass lookup: exact name match first, then alias fallback. Added `test_find_dependents_exact_name_before_alias` regression test to verify a project named `util` wins over a project that merely provides `util`.

File: src/lib.rs [hunk 11]
```diff
@@ -812,4 +944,200 @@ mod tests {
...
+        );
+
+        match result {
+            CommandResult::Message(msg) => {
+                let parsed: Vec<String> = serde_json::from_str(&msg).unwrap();
+                assert_eq!(parsed, vec!["b", "c"]);
+            }
+            _ => panic!("Expected Message result"),
+        }
+    }
```
[the remark this answers is not in the document]
Comment (mateodelnorte) [reply src/lib.rs:1181]: Fixed in 9950cbd. Added `test_project_dependents_json_flag_in_args` covering both `["a", "--json"]` and `["--json", "a"]` forms through `execute_command`.

**Comment by mateodelnorte** [comment]
## Review fixes pushed

| Commit | Comment | Fix |
|--------|---------|-----|
| 9950cbd | T10: Prefer exact name matches before alias matches (@coderabbitai) | Two-pass lookup in `find_dependents`: exact name first, then `provides` fallback. Added regression test `test_find_dependents_exact_name_before_alias`. |
| 9950cbd | T11: Cover CLI-style `--json` forms in tests (@coderabbitai) | Added `test_project_dependents_json_flag_in_args` covering `["a", "--json"]` and `["--json", "a"]` through `execute_command`. |

**Build**: clippy clean, all 26 tests passing (including 2 new tests).

**Confidence**: 100% — all review threads resolved, code reviewed across correctness, deduplication, performance, security, and test coverage dimensions.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

**Comment by mateodelnorte** [comment]
🤖 **Agent starting** (`run-pr-review-1774473707`): `pr-review` run dispatched.

File: src/lib.rs [hunk 12]
```diff
@@ -324,6 +329,133 @@ fn format_project_tree(nodes: &[ProjectTreeNode], output: &mut String, prefix: &
...
+    let options = &effective_options;
+
+    let project_name = match args.iter().find(|a| !a.starts_with('-')) {
+        Some(name) => name.clone(),
+        None => {
+            return CommandResult::ShowHelp(Some(
+                "Usage: meta project dependents <project-name>".to_string(),
+            ))
+        }
+    };
```
[the remark this answers is not in the document]
Comment (mateodelnorte) [reply src/lib.rs:383]: Fixed in cd71e83: arg parsing now properly skips known flags and their values (--depth N, --json, --recursive, etc.) and rejects multiple positional arguments.

File: src/lib.rs [hunk 13]
```diff
@@ -324,6 +329,133 @@ fn format_project_tree(nodes: &[ProjectTreeNode], output: &mut String, prefix: &
...
+            Ok(j) => j,
+            Err(e) => return CommandResult::Error(format!("Failed to serialize JSON: {e}")),
+        };
+        CommandResult::Message(json)
+    } else {
+        if dependents.is_empty() {
+            CommandResult::Message(format!("No projects depend on '{project_name}'."))
+        } else {
+            CommandResult::Message(dependents.join("\n"))
+        }
```
[the remark this answers is not in the document]
Comment (mateodelnorte) [reply src/lib.rs]: Fixed in cd71e83: find_dependents now returns Option<Vec<String>> — None when the project/alias is unknown, producing a clear error instead of the misleading "No projects depend on..." message.

**Comment by mateodelnorte** [comment]
## Review Fix Summary

| Commit | Comment | Fix |
|--------|---------|-----|
| cd71e83 | T12: Parse `<project-name>` after stripping option/value pairs (@coderabbitai) | Proper flag-aware arg parsing: skips `--json`, `--recursive`, `-r`, `--verbose`, `--parallel`, `--depth N`, `--depth=N`; rejects multiple positional args |
| cd71e83 | T13: Unknown targets should fail instead of "no dependents" (@coderabbitai) | `find_dependents` returns `Option<Vec<String>>`; `None` → clear `"Unknown project or alias: ..."` error |

### Build status
- **Clippy**: clean (zero warnings with `-D warnings`)
- **Tests**: 29/29 passing (3 new tests for T12/T13 coverage)

### New tests added
- `test_project_dependents_depth_flag_skipped` — verifies `--depth 1 lib_a` and `--depth=1 lib_a` correctly resolve `lib_a`
- `test_project_dependents_multiple_positionals_error` — verifies `a b` produces "exactly one" error
- `test_project_dependents_unknown_project_error` — verifies unknown project name produces clear error

### Confidence: 100%

All review threads resolved. No issues flagged for human review.
